crossbeam/dsl/value.py

```diff
@@ -76,63 +94,87 @@ def get_weight(self):
     """Returns this expression's weight, computed recursively."""
 
 
+class Variable(Value):
+  """A variable name."""
+
+  def __init__(self, name, values=None, weight=1, is_free=False):
+    if values is None:
+      values = [f'{self.__class__.__name__}_{name}']
+    super(Variable, self).__init__(values,
+                                   free_variables=[self] if is_free else [])
+    self.name = name
+    self._weight = weight
+
+  def tokenized_expression(self):
+    return [self.name]
+
+  def get_weight(self):
+    return self._weight
+
+
+class FreeVariable(Variable):
+  """A free variable name, which implies the outermost Value is a lambda."""
+
+  def __init__(self, name):
+    super(FreeVariable, self).__init__(name, is_free=True)
+
+
+@functools.cache
+def get_free_variable(i):
+  return FreeVariable(f'v{i + 1}')
+
+
+class BoundVariable(Variable):
+  """A bound variable name, bound to a lambda in a lambda-requiring function."""
+
+  def __init__(self, name):
+    super(BoundVariable, self).__init__(name)
+
+
+@functools.cache
+def get_bound_variable(i):
```

Comment:
minor: why the name index starts at 1 while i starts at 0? ---- no need to change anything as long as it is consistent everywhere.

Reply:
because that's how we described the variables in our discussions (v1, v2, u1, u2, etc). it should be consistent because this is the only place this logic happens

Reply:
sounds good! thanks